### PR TITLE
[iOS] - Add Skus-Internals WebUI

### PR DESCRIPTION
**Pull request by Brandon-T**
## Summary
- Export Skus-UI and add Skus-UI internals interface for iOS

Resolves https://github.com/brave/brave-browser/issues/37222

## Submitter Checklist:

- [x] I confirm that no [security/privacy review is needed](https://github.com/brave/brave-browser/wiki/Security-reviews) and no other type of reviews are needed, or that I have [requested](https://github.com/brave/reviews/issues/new/choose) them
- [x] There is a [ticket](https://github.com/brave/brave-browser/issues) for my issue
- [x] Used Github [auto-closing keywords](https://docs.github.com/en/github/managing-your-work-on-github/linking-a-pull-request-to-an-issue) in the PR description above
- [x] Wrote a good [PR/commit description](https://google.github.io/eng-practices/review/developer/cl-descriptions.html)
- [x] Squashed any review feedback or "fixup" commits before merge, so that history is a record of what happened in the repo, not your PR
- [x] Added appropriate labels (`QA/Yes` or `QA/No`; `release-notes/include` or `release-notes/exclude`; `OS/...`) to the associated issue
- [x] Checked the PR locally:
  * `npm run test -- brave_browser_tests`, `npm run test -- brave_unit_tests` [wiki](https://github.com/brave/brave-browser/wiki/Tests)
  * `npm run presubmit` [wiki](https://github.com/brave/brave-browser/wiki/Presubmit-checks), `npm run gn_check`, `npm run tslint`
- [x] Ran `git rebase master` (if needed)

## Reviewer Checklist:

- [ ] A security review [is not needed](https://github.com/brave/brave-browser/wiki/Security-reviews), or a link to one is included in the PR description
- [ ] New files have MPL-2.0 license header
- [ ] Adequate test coverage exists to prevent regressions
- [ ] Major classes, functions and non-trivial code blocks are well-commented
- [ ] Changes in component dependencies are properly reflected in `gn`
- [ ] Code follows the [style guide](https://chromium.googlesource.com/chromium/src/+/HEAD/styleguide/c++/c++.md)
- [ ] Test plan is specified in PR before merging

## After-merge Checklist:

- [ ] The associated issue milestone is set to the smallest version that the
  changes has landed on
- [ ] All relevant documentation has been updated, for instance:
  - [ ] https://github.com/brave/brave-browser/wiki/Deviations-from-Chromium-(features-we-disable-or-remove)
  - [ ] https://github.com/brave/brave-browser/wiki/Proxy-redirected-URLs
  - [ ] https://github.com/brave/brave-browser/wiki/Fingerprinting-Protections
  - [ ] https://github.com/brave/brave-browser/wiki/Brave%E2%80%99s-Use-of-Referral-Codes
  - [ ] https://github.com/brave/brave-browser/wiki/Web-Compatibility-Exceptions-in-Brave
  - [ ] https://github.com/brave/brave-browser/wiki/QA-Guide
  - [ ] https://github.com/brave/brave-browser/wiki/P3A

### CODE REVIEW

File: components/resources/BUILD.gn [hunk 1]
```diff
@@ -60,19 +60,21 @@ repack("resources") {
   deps = [ ":static_resources" ]
   sources = [ "$root_gen_dir/components/brave_components_static.pak" ]
 
+  deps += [ "//brave/components/skus/browser/resources:generated_resources" ]
+
+  sources += [ "$root_gen_dir/brave/components/skus/browser/resources/skus_internals_generated.pak" ]
+
```
Comment (kylehickinson) [inline components/resources/BUILD.gn:66]: These can just be added to the `deps`/`sources` declarations above, no need to append

File: ios/browser/api/ui/webui/skus/BUILD.gn
Comment (kylehickinson) [inline ios/browser/api/ui/webui/skus/BUILD.gn:1]: Nothing in this file is exported in the BraveCore public framework headers so shouldn't be in the `api` subfolder. Please move to `ios/browser` directly

File: ios/browser/api/ui/webui/skus/skus_internals_ui.mm [hunk 2]
```diff
@@ -0,0 +1,285 @@
...
+    }
+
+    // Set output with env like {skus:production: {...}}.
+    dict.Set(kv.first, std::move(order_dict_output));
+  }
+  return dict;
+}
+
+void SkusInternalsUI::ResetSkusState() {
+  // TODO: Ask iOS side to reset skus
```
Comment (kylehickinson) [inline ios/browser/api/ui/webui/skus/skus_internals_ui.mm:198]: Can we not just `<if expr="not is_ios">` the button in the actual WebUI html instead? Better to just not show a button that isn't working on iOS

File: ios/browser/api/ui/webui/skus/skus_internals_ui.mm [hunk 3]
```diff
@@ -0,0 +1,285 @@
...
+#include "ios/web/public/webui/web_ui_ios.h"
+#include "ios/web/public/webui/web_ui_ios_data_source.h"
+#include "ios/web/public/webui/web_ui_ios_message_handler.h"
+#include "ui/base/clipboard/clipboard_ios.h"
+#include "ui/base/clipboard/scoped_clipboard_writer.h"
+#include "ui/base/webui/resource_path.h"
+#include "ui/base/webui/web_ui_util.h"
+
+@interface UIView (Private)
+- (UIViewController*)parentController;
```
Comment (kylehickinson) [inline ios/browser/api/ui/webui/skus/skus_internals_ui.mm:41]: Lets do one of these two things:
1. Implement this as just a basic function instead of a category
2. Prefix the method name with `brave_` (e.g. `brave_parentController`)

File: ios/browser/api/ui/webui/skus/skus_internals_ui.mm [hunk 4]
```diff
@@ -0,0 +1,285 @@
...
+@end
+
+@implementation UIView (Private)
+- (UIViewController*)parentController {
+  UIResponder* nextResponder = [self nextResponder];
+  if ([nextResponder isKindOfClass:[UIViewController class]]) {
+    return (UIViewController*)nextResponder;
```
Comment (kylehickinson) [inline ios/browser/api/ui/webui/skus/skus_internals_ui.mm:48]: `static_cast`?

**Review by kylehickinson — CHANGES_REQUESTED** [review]

File: ios/browser/api/ui/webui/skus/skus_internals_ui.mm (hunk 2 above)
Comment (Brandon-T) [reply ios/browser/api/ui/webui/skus/skus_internals_ui.mm]: No longer needed. It works fine since I fixed the alerts not working.

File: browser/ui/webui/skus_internals_ui.cc [hunk 5]
```diff
@@ -94,25 +95,36 @@ void SkusInternalsUI::GetVpnState(GetVpnStateCallback callback) {
...
+  auto* profile = Profile::FromWebUI(web_ui());
+  if (!brave_vpn::IsBraveVPNEnabled(profile->GetPrefs())) {
+    base::Value::Dict empty;
+    dict.Set("Order", empty);
+  } else {
+    auto order_info = GetOrderInfo("vpn.");
+    order_info.Set(
+        "env", local_state_->GetString(brave_vpn::prefs::kBraveVPNEnvironment));
+    dict.Set("Order", order_info);
+  }
```
Comment (kylehickinson) [inline browser/ui/webui/skus_internals_ui.cc:107]: Is this change even needed? This part won't evaluate for iOS due to `#if BUILDFLAG(ENABLE_BRAVE_VPN)` check around it

File: ios/browser/api/web/ui/chrome_webview.mm [hunk 6]
```diff
@@ -130,4 +269,86 @@ - (void)webStateDestroyed:(web::WebState*)webState {
...
+- (void)webController:(ChromeWebViewController*)controller
+    runJavaScriptConfirmPanelWithMessage:(NSString*)message
+                                 pageURL:(NSURL*)URL
+                       completionHandler:(void (^)(BOOL))handler {
+  UIAlertController* alert =
+      [UIAlertController alertControllerWithTitle:nil
+                                          message:message
+                                   preferredStyle:UIAlertControllerStyleAlert];
+
+  [alert addAction:[UIAlertAction actionWithTitle:@"Ok"
```
Comment (kylehickinson) [inline ios/browser/api/web/ui/chrome_webview.mm:309]: nit: `Ok` should be `OK`, or we should pull a localized string (`IDS_OK` and `IDS_CANCEL` exist in brave's component strings) but not really that necessary

File: browser/ui/webui/skus_internals_ui.cc (hunk 5 above)
Comment (kylehickinson) [reply browser/ui/webui/skus_internals_ui.cc:107]: Nvm I see it was pulled from `GetVPNOrderInfo`

File: browser/ui/webui/skus_internals_ui.cc [hunk 7]
```diff
@@ -94,25 +95,36 @@ void SkusInternalsUI::GetVpnState(GetVpnStateCallback callback) {
...
+  std::string result;
+  base::JSONWriter::Write(dict, &result);
+  std::move(callback).Run(result);
+}
 
-  const auto& skus_state = local_state_->GetDict(skus::prefs::kSkusState);
+base::Value::Dict SkusInternalsUI::GetOrderInfo(
+    const std::string& location) const {
+  base::Value::Dict dict const auto& skus_state =
+      local_state_->GetDict(skus::prefs::kSkusState);
```
Comment (simonhong) [inline browser/ui/webui/skus_internals_ui.cc:127]:
```suggestion
  base::Value::Dict dict;
  const auto& skus_state =
      local_state_->GetDict(skus::prefs::kSkusState);
```

File: browser/ui/webui/skus_internals_ui.cc [hunk 8]
```diff
@@ -94,25 +95,36 @@ void SkusInternalsUI::GetVpnState(GetVpnStateCallback callback) {
...
-  dict.Set("Order", GetVPNOrderInfo());
+  auto* profile = Profile::FromWebUI(web_ui());
+  if (!brave_vpn::IsBraveVPNEnabled(profile->GetPrefs())) {
+    base::Value::Dict empty;
+    dict.Set("Order", empty);
+  } else {
+    auto order_info = GetOrderInfo("vpn.");
+    order_info.Set(
+        "env", local_state_->GetString(brave_vpn::prefs::kBraveVPNEnvironment));
+    dict.Set("Order", order_info);
```
Comment (simonhong) [inline browser/ui/webui/skus_internals_ui.cc:106]: ditto
```suggestion
    dict.Set("Order", std::move(order_info));
```

File: browser/ui/webui/skus_internals_ui.cc [hunk 9]
```diff
@@ -94,25 +95,36 @@ void SkusInternalsUI::GetVpnState(GetVpnStateCallback callback) {
 #if !BUILDFLAG(IS_ANDROID)
   dict.Set("Last connection error", GetLastVPNConnectionError());
 #endif
-  dict.Set("Order", GetVPNOrderInfo());
+  auto* profile = Profile::FromWebUI(web_ui());
+  if (!brave_vpn::IsBraveVPNEnabled(profile->GetPrefs())) {
+    base::Value::Dict empty;
+    dict.Set("Order", empty);
```
Comment (simonhong) [inline browser/ui/webui/skus_internals_ui.cc:101]: Dict should be rvalue.
```suggestion
    dict.Set("Order", base::Value::Dict());
```

**Review by kylehickinson — APPROVED** [review]

**Comment by kjozwiak** [comment]
Verification **PASSED** on `iPhone 11` running `iOS 17.4.1` using the following build(s):

```
1.66.62 Chromium: 123.0.6312.105 (Official Build) nightly (64-bit)
--- | ---
Revision | e8a70be9c002...
OS | iOS
```

**`Quick Note:`** Because we can't purchase VPN/Leo on `Nightly` via TF, this can't be really thoroughly checked until it we get RC builds via the `release` channel. For the Nightly verification, just ensured that loading `brave://skus-internals` doesn't cause any problems/crashes.

* ensured that you can load `brave://skus-internals` without issues
* ensured that `Reset SKUs state`, `Fetch VPN state`, `Fetch Leo state`  & `Fetch SKUs state` doesn't run into issues when blank
  * basically ensured that using those features when there's no SKUs information doesn't crash/cause issues

As mentioned above, once we get an RC via `release` with the above fix, we'll verify it more thoroughly via https://github.com/brave/brave-browser/issues/37222.

IMG_0233
